### PR TITLE
[refactor] Switch to using staging buffer for metal/vulkan/opengl

### PR DESCRIPTION
**Pull request by ailzhang**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #5299
* #5298
* __->__ #5297

ndarray host accesses

### CODE REVIEW

File: taichi/program/ndarray.cpp [hunk 1]
```diff
@@ -103,24 +119,96 @@ std::size_t Ndarray::get_nelement() const {
   return nelement_;
 }
 
+template <typename T>
+T Ndarray::read(const std::vector<int> &I) const {
+  prog_->synchronize();
+  size_t index = flatten_index(total_shape_, I);
+  size_t size_ = sizeof(T);
```
Comment (bobcao3) [inline taichi/program/ndarray.cpp:126]: How large would T typically be? If it's limited to primitive types, we can reuse the same buffer for these accessors to save the cost of buffer allocation

File: tests/cpp/program/graph_test.cpp [hunk 2]
```diff
@@ -48,6 +48,10 @@ TEST(GraphTest, SimpleGraphRun) {
   args.insert({"x", aot::IValue::create<int>(2)});
 
   g->run(args);
+  // TODO: this should be synchronize() indeed but it currently has a shortcut
+  // `sync=True` to return which is pretty error-prone. We should change this to
+  // synchronize() after getting rid of the shortcut.
```
Comment (bobcao3) [inline tests/cpp/program/graph_test.cpp:53]: ooc, why can't we change this to `synchronize` rn?

**Review by bobcao3 — APPROVED** [review]
LGTM, other than the one nitpicking

File: taichi/program/ndarray.cpp (hunk 1 above)
Comment (ailzhang) [reply taichi/program/ndarray.cpp:126]: Yup in this case it's typically primitive types, will send a followup PR to cache buffer allocation. :D

**Pull request merged**